File: core/handlers/store_account.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"strings"
	"time"

	"dylaris-core/services"
)

// The store account, read and changed from the panel.
//
// A tenant runs their servers here and pays over there, and until now the panel
// could only tell them WHETHER the two were joined. What they actually want to
// know - which subscription, how much of the traffic and backup allowance is
// gone, and whether going past it bills them or stops them - lived behind a
// store session the panel does not have.
//
// So Core asks on their behalf over the shared-key channel it already uses for
// link-status. Two rules make that safe, and both are load-bearing:
//
//   - The UUID is taken from the SESSION, never from the request body. A caller
//     who could name the account would be able to read anyone's invoice and move
//     anyone's consent.
//   - The store looks that UUID up in its own join table rather than trusting it
//     as an identity. Core proves it is Core with the shared key; the key says
//     nothing about which tenant.
//
// The numbers are not recomputed here. The store owns the money and already
// computes them for its own screens; a second implementation is the one that
// drifts, and the platform has been bitten by exactly that shape before - a
// pricing page promising an allowance nothing enforced.

// storeAccountTimeout bounds the round trip. The panel renders the rest of the
// page without this, so a slow storefront must degrade to "we could not ask"
// rather than hold the whole view.
const storeAccountTimeout = 8 * time.Second

// AccountSummary GET /api/store/account-summary - session-authed.
//
// Best-effort by design: a store outage returns the reachable=false shape with a
// reason, not an error page. The panel it is embedded in still works, and the
// tenant is told which of the two systems is quiet rather than being shown an
// account that appears to have nothing in it.
func (h *StoreHandler) AccountSummary(w http.ResponseWriter, r *http.Request) {
	if !h.state.StoreEnabled {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{"success": true, "enabled": false})
		return
	}
	userID, _ := r.Context().Value("userID").(string)
	if userID == "" {
		sendJSONError(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	summary, err := h.state.storeAccountSummary(r.Context(), userID)
	if err != nil {
		log.Printf("store: account summary for %s: %v", userID, err)
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"success":   true,
			"enabled":   true,
			"reachable": false,
			"message":   "The store could not be reached, so your subscription details are not shown. Nothing about your servers is affected.",
		})
		return
	}
	summary["success"] = true
	summary["enabled"] = true
	summary["reachable"] = true
	summary["storeUrl"] = h.state.StoreURL
	// Whether an ADMIN GRANT is what entitles them, which the storefront cannot
	// know: a grant made in the panel writes Core's billing row and creates no
	// store subscription at all.
	//
	// Without it the panel told a granted tenant "There is no active
	// subscription to bill traffic against" - true of the store's database and
	// unreadable to somebody whose grant is working, because it describes their
	// entitlement as missing when it is simply not a purchase.
	if ent, eerr := services.EffectiveEntitlement(h.state.Store, userID, time.Now(), h.state.StoreEnabled, IsAdmin(r)); eerr == nil {
		summary["granted"] = ent.Source == services.EntitlementSourceGrant || ent.Source == services.EntitlementSourceBoth
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(summary)
}

// SetBillingConsent POST /api/store/billing-consent - session-authed.
// Body {traffic, backup}; an omitted field is left alone.
//
// There is deliberately no uuid in that body. The one this acts on is the one
// the session authenticated, which is the whole reason this endpoint can exist
// in front of a shared-key channel that would otherwise let any caller name any
// account.
func (h *StoreHandler) SetBillingConsent(w http.ResponseWriter, r *http.Request) {
	if !h.state.StoreEnabled {
		sendJSONError(w, "The store integration is not configured", http.StatusNotFound)
		return
	}
	userID, _ := r.Context().Value("userID").(string)
	if userID == "" {
		sendJSONError(w, "Unauthorized", http.StatusUnauthorized)
		return
	}
	var req struct {
		Traffic *bool `json:"traffic"`
		Backup  *bool `json:"backup"`
	}
	if err := json.NewDecoder(io.LimitReader(r.Body, 1<<12)).Decode(&req); err != nil {
		sendJSONError(w, "Invalid input", http.StatusBadRequest)
		return
	}
	if req.Traffic == nil && req.Backup == nil {
		sendJSONError(w, "Nothing to change", http.StatusBadRequest)
		return
	}

	status, message, err := h.state.storeSetBillingConsent(r.Context(), userID, req.Traffic, req.Backup)
	if err != nil {
		log.Printf("store: billing consent for %s: %v", userID, err)
		sendJSONError(w, "The store could not be reached, so nothing was changed.", http.StatusBadGateway)
		return
	}
	if status != http.StatusOK {
		// The store's own words. It knows why - no subscription, no meter
		// configured, not billed through Stripe - and each of those sends the
		// reader somewhere different.
		sendJSONError(w, message, status)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{"success": true})
}

// storeAccountSummary fetches the tenant's picture from the storefront.
func (s *AppState) storeAccountSummary(ctx context.Context, uuid string) (map[string]interface{}, error) {
	endpoint := strings.TrimRight(s.StoreURL, "/") + "/api/store/account-summary?uuid=" + url.QueryEscape(uuid)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("X-Store-Key", s.StoreSharedKey)

	resp, err := (&http.Client{Timeout: storeAccountTimeout}).Do(req)
	if err != nil {
		return nil, fmt.Errorf("cannot reach the storefront at %s: %w", s.StoreURL, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
		return nil, fmt.Errorf("the storefront refused our key (HTTP %d) - STORE_SHARED_KEY does not match the value configured on %s", resp.StatusCode, s.StoreURL)
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("the storefront answered HTTP %d", resp.StatusCode)
	}
	body, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<18))
	var parsed map[string]interface{}
	if err := json.Unmarshal(body, &parsed); err != nil {
		return nil, fmt.Errorf("the storefront answered something that is not the expected JSON: %w", err)
	}
	return parsed, nil
}

// storeSetBillingConsent moves one or both switches for this tenant.
//
// It returns the store's status and message rather than collapsing them: a 409
// "no active subscription", a 503 "no meter configured" and a 502 "the payment
// provider said no" are three different people's problem, and one generic
// failure would send all three to the wrong place.
func (s *AppState) storeSetBillingConsent(ctx context.Context, uuid string, traffic, backup *bool) (int, string, error) {
	payload := map[string]interface{}{"uuid": uuid}
	if traffic != nil {
		payload["traffic"] = *traffic
	}
	if backup != nil {
		payload["backup"] = *backup
	}
	body, _ := json.Marshal(payload)

	endpoint := strings.TrimRight(s.StoreURL, "/") + "/api/store/billing-consent"
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, strings.NewReader(string(body)))
	if err != nil {
		return 0, "", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Store-Key", s.StoreSharedKey)

	resp, err := (&http.Client{Timeout: storeAccountTimeout}).Do(req)
	if err != nil {
		return 0, "", fmt.Errorf("cannot reach the storefront at %s: %w", s.StoreURL, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusOK {
		return http.StatusOK, "", nil
	}
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<16))
	var parsed struct {
		Error string `json:"error"`
	}
	_ = json.Unmarshal(raw, &parsed)
	if parsed.Error == "" {
		parsed.Error = fmt.Sprintf("The store refused the change (HTTP %d).", resp.StatusCode)
	}
	return resp.StatusCode, parsed.Error, nil
}

File: panel/src/lib/storeAccount.ts
import type { StoreAccountSummary } from '@/lib/api/store';

// Presentation rules for the store account card. Pure, so the ones that decide
// whether a switch can be moved and what an allowance bar says are testable
// without a storefront.

// usagePct is what the bar fills to.
//
// A ceiling of zero is NOT a full bar. It means the tenant may hold none of this
// at all, and drawing 100% would read as "you are at your limit" to somebody who
// never had one. Anything at or over the ceiling is 100 - a bar that keeps
// growing past its track is a bar that has stopped meaning anything.
export function usagePct(used: number, ceiling: number): number {
    if (!Number.isFinite(used) || !Number.isFinite(ceiling) || ceiling <= 0) return 0;
    return Math.max(0, Math.min(100, Math.round((used / ceiling) * 100)));
}

export type SwitchState =
    | { kind: 'on' }
    | { kind: 'off' }
    | { kind: 'unavailable'; reason: string };

// trafficSwitch and backupSwitch decide whether the tenant may move the control
// at all, and say why when they may not.
//
// Three different "no"s, and collapsing them is how somebody spends an afternoon
// on the wrong problem: nothing to bill against, no meter configured by the
// operator, or a subscription that is not billed through Stripe at all. Only the
// second is anybody's fault, and it is not the tenant's.
function billingSwitch(
    s: StoreAccountSummary,
    enabled: boolean | undefined,
    meterConfigured: boolean | undefined,
    what: string,
): SwitchState {
    if (!s.subscribed) {
        // A grant is not a missing subscription, and saying so was actively
        // misleading: a tenant whose BYON works read "there is no active
        // subscription" as though their access were broken. A grant made in the
        // panel writes Core's billing row and creates no store subscription at
        // all, so there is nothing to meter - and nothing sweeps them either,
        // because the guard only walks Stripe subscriptions.
        if (s.granted) {
            return {
                kind: 'unavailable',
                reason: `Your access was granted by an administrator rather than bought, so there is no subscription to meter ${what} against. Nothing is charged, and nothing is stopped for going over.`,
            };
        }
        return { kind: 'unavailable', reason: `There is no active subscription to bill ${what} against.` };
    }
    if (!s.stripe) {
        return {
            kind: 'unavailable',
            reason: 'This subscription was granted by an administrator rather than billed through the store, so there is nothing to meter.',
        };
    }
    if (!meterConfigured) {
        return {
            kind: 'unavailable',
            reason: `Metered ${what} is not set up on the store yet. Nothing you can do from here.`,
        };
    }
    return enabled ? { kind: 'on' } : { kind: 'off' };
}

export function trafficSwitch(s: StoreAccountSummary): SwitchState {
    return billingSwitch(s, s.trafficBillingEnabled, s.trafficMeterConfigured, 'traffic');
}

export function backupSwitch(s: StoreAccountSummary): SwitchState {
    return billingSwitch(s, s.backupBillingEnabled, s.backupMeterConfigured, 'backup storage');
}

// consequenceOf is the sentence under each switch: what happens when the
// allowance runs out, in the state the switch is currently in.
//
// Written as a consequence rather than as a setting name because that is the
// decision being made. "Metered billing: off" tells a tenant nothing; "you will
// be stopped rather than charged" tells them exactly what they are choosing.
export function consequenceOf(kind: 'traffic' | 'backup', on: boolean): string {
    if (kind === 'traffic') {
        return on
            ? 'Past your included traffic you keep running and the extra is billed by the terabyte.'
            : 'Past your included traffic your servers stop being reachable until the next period. Nothing is charged.';
    }
    return on
        ? 'Past your included backup storage new backups keep running and the extra is billed per block.'
        : 'Past your included backup storage new backups are refused. Nothing is charged, and nothing already stored is deleted.';
}
